backend.native/tests/runtime/concurrent/shared1.kt

```diff
@@ -0,0 +1,508 @@
+
+    localRef.value = null
+    GC.collect()
+
+    assertNull(localWeak.get())
+    assertNull(localValueWeak.get())
+}
+
+@Test fun testDisposeAndCollect() {
+    val (localRef, localWeak, localValueWeak) = getWeaksAndAtomicReference(3)
+
+    localRef.value!!.dispose()
+    GC.collect()
+
+    assertNotNull(localWeak.get())
+    assertNull(localValueWeak.get())
+}
+
+fun collectInWorker(worker: Worker, semaphore: AtomicInt): Pair<WeakReference<A>, Future<Unit>> {
+    val (localRef, _, localValueWeak) = getWeaksAndAtomicReference(3)
+
+    val future = worker.execute(TransferMode.SAFE, { Pair(localRef, semaphore) }) { (localRef, semaphore) ->
+        semaphore.increment()
+        while (semaphore.value < 2) {}
+
+        localRef.value = null
+        GC.collect()
+    }
+
+    while (semaphore.value < 1) {}
+    GC.collect()
+    assertNotNull(localValueWeak.get())
+
+    return Pair(localValueWeak, future)
+}
+
+@Test fun testCollectInWorker() {
+    val semaphore: AtomicInt = AtomicInt(0)
+
+    val worker = Worker.start()
+
+    val (localValueWeak, future) = collectInWorker(worker, semaphore)
+    semaphore.increment()
+
+    future.result
+    GC.collect()
+    assertNull(localValueWeak.get())
+    worker.requestTermination().result
+}
+
+fun doNotCollectInWorker(worker: Worker, semaphore: AtomicInt): Future<DisposableSharedRef<A>> {
+    val local = DisposableSharedRef.create(A(3))
+
+    return worker.execute(TransferMode.SAFE, { Pair(local, semaphore) }) { (local, semaphore) ->
+        semaphore.increment()
+        while (semaphore.value < 2) {}
+
+        GC.collect()
+        local
+    }
+}
+
+@Test fun testDoNotCollectInWorker() {
+    val semaphore: AtomicInt = AtomicInt(0)
+
+    val worker = Worker.start()
+
+    val future = doNotCollectInWorker(worker, semaphore)
+    while (semaphore.value < 1) {}
+    GC.collect()
+    semaphore.increment()
+
+    val value = future.result
+    assertEquals(3, value.get().a)
+    worker.requestTermination().result
+}
+
+fun disposeInWorker(worker: Worker, semaphore: AtomicInt): Triple<WeakReference<DisposableSharedRef<A>>, WeakReference<A>, Future<Unit>> {
+    val (localRef, localWeak, localValueWeak) = getWeaksAndAtomicReference(3)
+
+    val future = worker.execute(TransferMode.SAFE, { Pair(localRef, semaphore) }) { (localRef, semaphore) ->
+        semaphore.increment()
+        while (semaphore.value < 2) {}
+
+        localRef.value!!.dispose()
+        GC.collect()
+    }
+
+    while (semaphore.value < 1) {}
+    GC.collect()
+    assertNotNull(localValueWeak.get())
+
+    return Triple(localWeak, localValueWeak, future)
+}
+
+@Test fun testDisposeInWorker() {
+    val semaphore: AtomicInt = AtomicInt(0)
+
+    val worker = Worker.start()
+
+    val (localWeak, localValueWeak, future) = disposeInWorker(worker, semaphore)
+    semaphore.increment()
+
+    future.result
+    GC.collect()
+    assertNotNull(localWeak.get())
```

Comment:
Why is it so?

Reply:
Left a comment explaining the reasoning, but will work on naming.

Reply:
I still don't get it.
Is `localWeak.get()` a `DisposableSharedRef`? What exactly prevents it from removing by `GC.collect()` above?

Reply:
Right, fixed it by holding onto `AtomicReference` of `DisposableSharedRef` on the main thread